lib/galaxy/tools/evaluation.py

```diff
@@ -543,9 +545,16 @@ def __build_environment_variables(self):
             directory = self.local_working_directory
             environment_variable = environment_variable_def.copy()
             environment_variable_template = environment_variable_def["template"]
+            inject = environment_variable_def.get("inject")
+            if inject == "api_key":
+                from galaxy.managers import api_keys
+                environment_variable_template = api_keys.ApiKeyManager(self.app).get_or_create_api_key(self._user)
```

Comment:
self._user would be `None` for anonymous users, failing with:
```
Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/lib/galaxy/jobs/runners/__init__.py", line 230, in prepare_job
    job_wrapper.prepare()
  File "/Users/mvandenb/src/galaxy/lib/galaxy/jobs/__init__.py", line 1066, in prepare
    self.command_line, self.extra_filenames, self.environment_variables = tool_evaluator.build()
  File "/Users/mvandenb/src/galaxy/lib/galaxy/tools/evaluation.py", line 484, in build
    raise e
  File "/Users/mvandenb/src/galaxy/lib/galaxy/tools/evaluation.py", line 481, in build
    self.__build_environment_variables()
  File "/Users/mvandenb/src/galaxy/lib/galaxy/tools/evaluation.py", line 551, in __build_environment_variables
    environment_variable_template = api_keys.ApiKeyManager(self.app).get_or_create_api_key(self._user)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/managers/api_keys.py", line 22, in get_or_create_api_key
    if user.api_keys:
AttributeError: 'NoneType' object has no attribute 'api_keys'
```
Maybe we can raise a more explicit exception in the `_user` property ?

Reply:
Maybe I'll just set the variable to empty - if the tool author wants to require a user they should use ``required_login`` so presumably if an anonymous user is using this tool the API key is optional?

cc @bgruening

Reply:
Agreed, this seems logical. Thanks!